### PR TITLE
Stack allocate on and task function argument bundles

### PR DESCRIPTION
**Pull request by mppf**
We've been heap allocating argument bundles for task bodies and for on statements.
This may have once been necessary but it does not seem to be necessary any longer.

For on statements:
- GASNet already has a model, for AMRequestMedium (which is what we eventually call to do an on statement), in which the passed buffer can be reused immediately after that GASNet function returns. Additionally, the buffer need not be in communication registered memory.
- Generally, we are trying to get to task stacks being allocated from communication-registered memory anyway. Additionally, the communication layer is free to copy into a heap allocated area if necessary. Lastly, on "on" statement causes the arguments to be copied over the network, so it is a fiction that by heap allocating them we preserve any sense of identity.

For tasks:
- tasks-qthreads.c already calls qthread_fork_copyargs and more generally qthread_spawn copies arguments to a thread (and heap allocates in some case if they are too big).
- for FIFO, the tasking layer was already calling the allocator for each spawned task in order to create a new task pool entry. This allocation can easily be extended to include space for the argument bundle.
- Any tasking layer that does not have the capability to copy some variable-length arguments to the task can heap-allocate and copy the arguments.

One issue with no longer heap-allocating the argument bundles is that the tasking and communication layer need to pass some additional information to the remote locale / executing task. For example, comm-gasnet.c needs to pass the requesting nodeID in order to handle execute_on_nb. Since the network communication will copy a single buffer, if the size of this comm-layer information was private, the comm layer would have to heap or stack allocate a new the bundle (that it includes the extra information) and copy in the old bundle's information. At the same time, since the Chapel comm layer interface is for the code generated by the compiler (vs a user-facing API), there is not much encapsulation benefit to making the size of the comm-layer information private in this way. Therefore this PR uses the alternative strategy, which is to allocate all argument bundles with a runtime header field already included. The runtime uses this space as needed in order to pass data (like the requesting nodeID). This extra field is included as the first field in the argument bundle class. The entire argument bundle is passed to on/task wrapper functions and this extra field is ignored at that point.

Note that in some cases a communication layer might prefer to copy the argument bundle into a payload of a different sort of message with a different sort of header. In such cases, the original argument bundle can be formed again in a stack variable on the other side. I expect that such an approach is fine for small message sizes.

Because tasks creation and execute_on require different runtime headers, the generated code differentiates between these cases. Since execute_on and execute_on_nb are generally implemented by creating tasks, the execute_on runtime header includes the tasking runtime header as its first element. In this way, the task header from any argument bundle can be read by the runtime regardless of the type.

Here is a tour through the changes in this PR:
- Adds _PRIM_STACK_ALLOCATE_CLASS_ primitive and adjusts codegenPrimitive to support it. This primitive will create stack space for a class instance. Since the Chapel compiler normally heap allocates class instances, the class instance type is a pointer type. I considered an alternative of storing argument bundles as classes containing a record, but that seemed to be too awkward a strategy. Also, this primitive might be useful in other cases. believe that converting heap allocations to stack allocations is an interesting optimization and this primitive might be able to help with that case.
- Adjusts bundleArgs in the compiler to call PRIM_STACK_ALLOCATE_CLASS.
- Adds an _args_size_ argument to runtime task and execute_on interfaces and adjust codegenInvokeTaskFun and codegenInvokeOnFun accordingly. This args_size argument is critical to enabling the runtime to copy the argument bundle.
- Changes the args argument to runtime task and execute_on functions to be the pointer type for the runtime header. The generated code puts this runtime header as the first element of the argument bundle struct and further arguments are in following structure members. There are actually two runtime headers: _chpl_comm_on_bundle_t_ and _chpl_task_bundle_t_. Adjusts the compiler to become aware of these types and to store them as the first element of the argument bundles.
- Adjusts comm-none.c, comm-gasnet.c, tasks-fifo.c, and tasks-qthreads.c to support the new interface and to be sure to copy the argument bundles before the relevant runtime call returns. Generally take care to minimize the number of allocations.

More about the format of the runtime headers:
- chpl-comm.h declares  chpl_comm_on_bundle_t and each task layer implementation declares chpl_task_bundle_t. The chpl_comm_on_bundle has a struct member, comm, that is defined by the communications layer. The comm layers define this in chpl-comm-task-decls.h.
- The bundle chpl_task_bundle_t is implementation private. The chpl_comm_on_bundle_t  is also largely implementation private but it contains a chpl_task_bundle_t and also a communication-layer private field.

### Performance results

Initial Performance Results:
- miniMD with compopts 1 or explicit version are 1.2x faster; other versions show less benefit.
- release lulesh ran 1.10x or 1.13x faster depending on configuration.

(current) SSCA2.4 gasnet+qthreads, on an XC
 * 1 node, stack-alloc  0.176849 vs master 0.208189 s
 * 4 node, stack-alloc  2.78988 vs master 2.77691 s
 * 16 node, stack-alloc 1.48196 vs master 1.49322 s

See [perf playground results](http://chapel.sourceforge.net/perf/chapcs/stackArgs/?startdate=2016/11/26&enddate=2016/12/03&configs=default,stackargs&graphs=lcalsrawompmedium,lcalsrawomplong,luleshrelease,luleshbradcstudyversion,minimdljsize10time,nasparallelbenchmarksfttimingssizes,nasparallelbenchmarksfttimingssizew,nasparallelbenchmarksfttimingssizea,emptytaskspawntimings500000xmaxtaskpar,reductionstimesec,timetobuildtrees)

### Testing

- bundles.chpl passes in all configurations I can think of where hello works on master
- bundles.chpl does not leak memory with quickstart,  or qthread+gasnet
- hellos work with default+NUMA configuration
- hellos work with --llvm
- hellos work with --baseline and/or --verify
- hellos, bundles.chpl work on Mac OS X with GASNet and segment=fast
- full test suite with comm=gasnet and tasks=qthreads
- full local testing (qthreads)
- full local quickstart testing

Reviewed by @gbtitus and @ronawho - thanks!

### CODE REVIEW

**Comment by mppf** [comment]
@ronawho @gbtitus I think you'll be interested in this PR to stack allocate arg bundles.

**Comment by ronawho** [comment]
I'm interested, but I definitely won't get to it today, and I'm out on Monday.

**Comment by gbtitus** [comment]
Starting to look at this now.  @ronawho , I'm not going to look very hard at the compiler changes, more at the runtime and module code.

File: runtime/include/chpl-align.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+#include <assert.h>
+#include <stdint.h>
+#include <limits.h>
+
+// If we have a mask representing 2^n - 1,
+// round an offset down to a multiple of 2^n.
+static inline
+uintptr_t round_down_to_mask(uintptr_t p, uintptr_t mask)
+{
+ return p - (p & mask);
```
Comment (gbtitus) [inline runtime/include/chpl-align.h]: one more space here

File: runtime/include/chpl-align.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  return (unsigned char*) round_down_to_mask((uintptr_t)p, mask);
+}
+
+// If we have a mask representing 2^n - 1,
+// round an offset up to a multiple of 2^n.
+static inline
+uintptr_t round_up_to_mask(uintptr_t p, uintptr_t mask)
+{
+  uintptr_t offset = p & mask;
+  return (offset==0)?(p):(p + (mask+1) - offset);
```
Comment (gbtitus) [inline runtime/include/chpl-align.h]: `return (p + mask) & ~mask;` avoids the control flow.

File: runtime/include/chpl-comm.h [hunk 3]
```diff
@@ -77,6 +77,44 @@ extern const int chpl_heterogeneous;
...
+}
+
+//
+// Call a chpl_ftable[] function in a task.
+//
+// This is a convenience function for use by the module code, in which
+// we have function table indices rather than function pointers.
+//
+static inline
+void chpl_task_taskCallFTable(chpl_fn_int_t fid,      // ftable[] entry to call
```
Comment (gbtitus) [inline runtime/include/chpl-comm.h]: Let's not define chpl_task_*() functions in chpl-comm.h.  Instead, leave chpl_task_taskCallFTable() where it was in chpl-tasks.h and define it as taking a chpl_task_bundle_p.  Add a new chpl_comm_taskCallFtable() that takes a chpl_comm_on_bundle_p here, call it from the locale model module code, and have it in turn call chpl_task_taskCallFTable() with the chpl_comm_on_bundle_task_bundle().  If everything is static inline it'll all collapse down to the same amount of code, but it'll be easier to find everything.

File: runtime/include/chpl-gen-includes.h [hunk 4]
```diff
@@ -35,9 +35,9 @@
 // one argument.
 //
 static inline
-void chpl_ftable_call(chpl_fn_int_t fid, void* arg)
+void chpl_ftable_call(chpl_fn_int_t fid, chpl_comm_on_bundle_t* bundle)
```
Comment (gbtitus) [inline runtime/include/chpl-gen-includes.h]: Interesting!  I would have bet we had calls to chpl_ftable_call() in both the comm and tasking layers, but indeed not; all of them are done from the comm layer.  I can imagine a future in which locale model module code calls this directly for local tasks, but that may never come to pass, so this change is fine as it stands.  I.e., this is just a comment, not a request for a change in the PR.

File: runtime/include/chpl-tasks.h [hunk 5]
```diff
@@ -45,6 +44,33 @@
...
+  chpl_bool is_executeOn;
+  c_sublocid_t requestedSubloc;
+  chpl_fn_p requested_fn;
+
+  chpl_task_bundleData_t task_prv;
+} chpl_task_bundle_t;
+
+typedef chpl_task_bundle_t *chpl_task_bundle_p;
+
+/* This shouldn't be needed because the task function
```
Comment (gbtitus) [inline runtime/include/chpl-tasks.h]: Remove commented-out func.

File: runtime/include/chpl-tasks.h [hunk 6]
```diff
@@ -104,6 +130,9 @@ void chpl_task_exit(void);        // called by the main task
 // thread) in order to be responsive and not be held up by other
 // user-level tasks. returns 0 on success, nonzero on failure.
 //
+// The caller of this function is responsible for ensuring that
+// *arg remains available to the task as long as it is running.
```
Comment (gbtitus) [inline runtime/include/chpl-tasks.h]: You might say "... as long as it is needed" here instead.  In actuality, both comm=gasnet and comm=ugni send a NULL for the arg and don't use it at all.

File: runtime/include/tasks/qthreads/tasks-qthreads.h [hunk 7]
```diff
@@ -117,6 +117,16 @@ extern
 volatile int chpl_qthread_done_initializing;
 
 typedef struct {
+  //chpl_bool serial_state;
```
Comment (gbtitus) [inline runtime/include/tasks/qthreads/tasks-qthreads.h]: Remove commented-out members.  But, it seems worthwhile to move the lineno and filename into the common code.  All the layer implementations need these two, after all.

File: runtime/src/chpl-cache-support.c [hunk 8]
```diff
@@ -56,33 +57,6 @@ static inline void saturating_add(saturating_count_t *x, unsigned int num) {
   *x = v;
 }
 */
-// If we have a mask representing 2^n - 1,
```
Comment (gbtitus) [inline runtime/src/chpl-cache-support.c]: Is this actually a related change (to the arg bundle stuff)?  Or is it an unrelated change that's gotten swept up in your branch?

File: runtime/include/tasks/qthreads/tasks-qthreads.h (hunk 7 above)
Comment (mppf) [inline runtime/include/tasks/qthreads/tasks-qthreads.h]: Actually, fifo doesn't need them. I don't think it'd be wrong to move them to the common code anyway...

File: runtime/src/chpl-cache-support.c (hunk 8 above)
Comment (mppf) [inline runtime/src/chpl-cache-support.c]: Earlier versions of the branch used round_up_to_mask, so I moved that function to a different header. I could remove it from the current version and possibly do it in a separate PR.

File: runtime/src/tasks/fifo/tasks-fifo.c [hunk 9]
```diff
@@ -721,7 +717,13 @@ void chpl_task_executeTasksInList(void** p_task_list_void) {
                            child_ptask->id,
                            child_ptask->is_executeOn);
 
-    (*task_to_run_fun)(child_ptask->arg);
+    if (child_ptask->countRunningTasks)
```
Comment (gbtitus) [inline runtime/src/tasks/fifo/tasks-fifo.c]: Don't believe it's correct to increment the running task count here. Conceptually, the parent task (which is what calls this function) is suspended while the children run.  In order to reduce overhead, and with the understanding that all task counting is inherently fuzzy anyway, we thus didn't bother to decrement the count while the parent is waiting, nor to increment it as the parent's thread is used here to execute children.  Is it important for the arg bundle change that the counting adjustment you've coded here be included, or is this just something you thought was missing and so is a "while I was here" change?

Comment (mppf) [inline runtime/src/tasks/fifo/tasks-fifo.c]: It's not a "while I was here" change. I moved the logic to adjust the counters from a wrapper task function (taskCallWrapper) into the main scheduler, so that I did not have to concern myself with further bundling/wrapping. Note that the current version in the branch with FIFO doesn't actually use the runtime area available for the task layer part of the bundle.

Comment (mppf) [inline runtime/src/tasks/fifo/tasks-fifo.c]: It only increments the task counts if it was requested to do so (it's conditional) so you'd have to look at what creates these ptasks and which set countRunningTasks=true. I believe that is the same as it used to be (just that it used to be done in callTaskWrapper).

Comment (gbtitus) [inline runtime/src/tasks/fifo/tasks-fifo.c]: Yeah, now I see what you've done.  This looks fine.

File: runtime/src/tasks/fifo/tasks-fifo.c [hunk 10]
```diff
@@ -1372,26 +1353,38 @@ static void maybe_add_thread(void) {
...
+
+
+  size_t payload_size;
+  task_pool_p ptask;
+
+  assert(a_size >= sizeof(chpl_task_bundle_t));
+
+  payload_size = a_size - sizeof(chpl_task_bundle_t);
+  ptask = (task_pool_p) chpl_mem_alloc(sizeof(task_pool_t) + payload_size,
+                                       //CHPL_RT_MD_TASK_POOL_DESC,
```
Comment (gbtitus) [inline runtime/src/tasks/fifo/tasks-fifo.c]: *ptask is now both a TASK_POOL_DESC memory and TASK_ARG memory.  We should invent a new CHPL_RT_MD value to reflect this.  The change here makes it different, but no more accurate.

File: runtime/src/tasks/massivethreads/tasks-massivethreads.c [hunk 11]
```diff
@@ -66,17 +61,19 @@ static int tasking_layer_active = 0;
 static int worker_in_cs_beforeinit = 0;
 static thread_local_data* s_tld;
 
-static const chpl_task_prvDataImpl_t s_def_chpl_data=
+// task-private data of comm task
```
Comment (gbtitus) [inline runtime/src/tasks/massivethreads/tasks-massivethreads.c]: My recollection is that tasks-massivethreads.c already didn't work, due to other tasking interface changes not having been done.  Does it build and test out for you?

Comment (mppf) [inline runtime/src/tasks/massivethreads/tasks-massivethreads.c]: it does now, for single locale

File: runtime/src/tasks/qthreads/tasks-qthreads.c [hunk 12]
```diff
@@ -732,33 +731,70 @@ static inline void wrap_callbacks(chpl_task_cb_event_kind_t event_kind,
     }
 }
 
-static aligned_t chapel_wrapper(void *arg)
+// If we stored chpl_taskID_t in chpl_task_bundleData_t,
+// this struct and the following function may not be necessary.
+typedef void (*main_ptr_t)(void);
+typedef struct {
+  chpl_task_bundle_t arg;
```
Comment (gbtitus) [inline runtime/src/tasks/qthreads/tasks-qthreads.c]: We can get away with this not being the last thing in the bundle here because chpl_main() doesn't take an argument, right?  If it did, then the need to have that argument directly follow the fixed fields in `arg` would mean that `arg` would have to be last?

File: runtime/src/tasks/qthreads/tasks-qthreads.c [hunk 13]
```diff
@@ -822,40 +865,47 @@ int chpl_task_createCommTask(chpl_fn_p fn,
...
+        arg->requestedSubloc   = subloc;
+        arg->requested_fn      = requested_fn;
+        arg->task_prv.lineno   = lineno;
+        arg->task_prv.filename = filename;
+
+        // wrap_callbacks assigns the task id on create?
+        // but that seems nasty. Also, we no longer need
+        // to form a chpl_task_prvDataImpl_t at this point.
+        // Can we call the event inside the task once it runs?
+        // wrap_callbacks(chpl_task_cb_event_kind_create,
```
Comment (gbtitus) [inline runtime/src/tasks/qthreads/tasks-qthreads.c]: The chpl_task_cb_event_kind_create event is supposed to occur as soon as the tasking layer knows about the task, basically, as soon as there's a descriptor for it.  The "begin" event is the one that occurs when the task starts running, but that's different.  So, this event needs to stay here.

File: runtime/src/tasks/qthreads/tasks-qthreads.c (hunk 12 above)
Comment (mppf) [inline runtime/src/tasks/qthreads/tasks-qthreads.c]: The argument doesn't actually have to follow the fixed fields, that was something I was predicting we would need but didn't end up needing.
It turns out that all of our on/task bodies know the argument bundle type they are starting with. So in this case that means that main_wrapper can just cast its argument to main_wrapper_bundle_t*. We need chpl_task_bundle_t to be the first structure member though so that the runtime can access those fields.

File: runtime/src/tasks/qthreads/tasks-qthreads.c [hunk 14]
```diff
@@ -865,47 +915,50 @@ void chpl_task_executeTasksInList(void **task_list)
...
+    bundle->requestedSubloc    = subloc;
+    bundle->requested_fn       = fp;
+    bundle->task_prv.lineno    = lineno;
+    bundle->task_prv.filename  = filename;
+
+    // wrap_callbacks assigns the task id on create?
+    // but that seems nasty. Also, we no longer need
+    // to form a chpl_task_prvDataImpl_t at this point.
+    // Can we call the event inside the task once it runs?
+    // wrap_callbacks(chpl_task_cb_event_kind_create,
```
Comment (gbtitus) [inline runtime/src/tasks/qthreads/tasks-qthreads.c]: Same comment with regard to the other "create" event.

File: runtime/src/tasks/qthreads/tasks-qthreads.c (hunk 13 above)
Comment (mppf) [inline runtime/src/tasks/qthreads/tasks-qthreads.c]: does chpl_task_cb_event_kind really need the final task ID?

Comment (gbtitus) [inline runtime/src/tasks/qthreads/tasks-qthreads.c]: I believe it does, in order to allow chplvis to figure out the parent/child relationships between tasks.  But check with Phil to be sure.

**Comment by gbtitus** [comment]
I'm through my first pass and although I made a number of comments I think this is looking pretty good. Obviously there's a ton of testing to do (including leak checks, valgrind, and chplvis testing), but this should yield quite nice performance benefits.

File: runtime/src/tasks/qthreads/tasks-qthreads.c (hunk 13 above)
Comment (mppf) [inline runtime/src/tasks/qthreads/tasks-qthreads.c]: Generally speaking, I wouldn't expect a task/thread system to assign a task descriptor until the task starts. E.g. pthreads won't give you a pthread id until you launch the thread. I'm not sure we want to require our tasking implementations to get a task ID before the task starts. I'll ask Phil about it.

Comment (gbtitus) [inline runtime/src/tasks/qthreads/tasks-qthreads.c]: Pthreads assign a unique ID to the child pthread at the time the child is created.  That ID is the value returned in the variable to which the first argument of pthread_create() points.  The parent can see this as soon as pthread_create() returns, which may be an arbitrarily long time before the child actually starts running.

Comment (mppf) [inline runtime/src/tasks/qthreads/tasks-qthreads.c]: Ah, you are right about pthreads. OK.

**Comment by mppf** [comment]
@gbtitus - did you look at the questions I had in the PR message? Also, I think I could go further and merge the task private data with the chpl_task_bundle_t if that is an appealing change to you.

File: runtime/include/tasks/qthreads/tasks-qthreads.h (hunk 7 above)
Comment (mppf) [inline runtime/include/tasks/qthreads/tasks-qthreads.h]: I'd still like to double-check. Do you think it's worth moving lineno and filename to the common bundle area, even though all tasking layers do not need them?

**Comment by mppf** [comment]
@gbtitus - I believe I've addressed all of your feedback. I still have some questions for you:

* Do you think it's worth moving lineno and filename to the common bundle area, even though all tasking layers do not use them?
* Should I attempt to merge the task-local storage with the argument bundle, since tasks can use the argument bundle area for this anyway? In particular, in qthreads we could just set the task local storage to be a pointer to the argument bundle...
* We could more radically change chpl-comm.h and chpl-tasks.h. In particular, many of the arguments to these functions (serial_state, requested function, line, file) could be stored in the argument bundle directly. I did not attempt to do so because I don't see a clear advantage but in some ways it would be simpler. Instead, the current implementation sets these elements of the argument bundle within the task/comm runtime layer call. Is that worth trying to do, or not useful? (We could also set these elements of the argument bundle in code common to the tasking/comm layers but leave the compiler/module's view of the runtime API the same).

**Comment by gbtitus** [comment]
Answers to above questions:
- Yes, I would move lineno and filename to the common bundle area.  Though not all layer implementations (currently) use those, arguably they ought to, and the symmetry is appealing.
- I think merging the task-local storage with the arg bundle (and having the qthreads qthread private data pointer point there) is a good idea.  There's a lot of overlap between the two anyway, and doesn't it save some copying (perhaps within Qthreads) at qthread/task creation time?
- Moving serial state etc. from the module/runtime calling sequence to the arg bundle could be done in a separate PR, later, if we decide it's a better way to go.  Let's leave that change out for now.

**Comment by mppf** [comment]
I'm going to try merging the arg bundle with the task-local-storage. It will slightly further increase the size of the arg bundle (by 8 bytes). That comes at some minor cost to communication performance (since the minimum message size is larger, since the task bundle is included in on statement AMs). But, we avoid some copying... Anyway, I'll give it a try but leave it a choice the tasking layer implementation can make.

**Comment by mppf** [comment]
@gbtitus - the change I just pushed gets rid of the duplication between argument bundle and task local storage but leaves a few things in task local storage only. That way, we don't have to pay the cost to move them for an executeOn. I did this only for qthreads so far, but I can do it also for fifo.

**Comment by gbtitus** [comment]
This seems fine.  I don't think the extra 8 bytes of transfer will be noticed; executeOn performance is limited by network latency (on the order of 1 usec per executeOn on an Aries network, not network bandwidth (up to on the order of 10 GB/s on an Aries network).

File: runtime/include/tasks/qthreads/tasks-qthreads.h [hunk 15]
```diff
@@ -118,33 +118,24 @@ volatile int chpl_qthread_done_initializing;
...
 typedef struct chpl_qthread_tls_s {
-    /* Task private data: serial state, etc. */
-    chpl_task_prvDataImpl_t chpl_data;
-    /* Reports */
-    int32_t    lock_filename;
-    size_t      lock_lineno;
+  chpl_task_bundle_p bundle;
+  // The below fields could move to chpl_task_bundleData_t
+  // That would reduce the size of the task local storage,
+  // but increase the size of executeOn bundles.
```
Comment (gbtitus) [inline runtime/include/tasks/qthreads/tasks-qthreads.h]: What would be the downside of increasing the size of the executeOn bundles by these two bytes?  It doesn't seem like it would add significantly to the stack space cost, nor to the executeOn network overhead (because dominated by latency rather than bandwidth).

Comment (mppf) [inline runtime/include/tasks/qthreads/tasks-qthreads.h]: It would increase it by 16 bytes, but yeah, it's unlikely to be significant. But, it seems unnecessary and I don't think it makes the code clearer or otherwise improves much, now that the duplicated fields are removed.

Comment (mppf) [inline runtime/include/tasks/qthreads/tasks-qthreads.h]: So in other words, you said "What is the downside" to which I responded "What is the upside?"

**Comment by ronawho** [comment]
I don't have any comments for the compiler changes, and if Greg's happy with the runtime changes, that's good enough for me

File: runtime/include/tasks/qthreads/tasks-qthreads.h (hunk 15 above)
Comment (gbtitus) [inline runtime/include/tasks/qthreads/tasks-qthreads.h]: It seems to produce a nice simplification -- doesn't moving lock_lineno and lock_filename into chpl_task_bundleData_t allow us to get ride of chpl_qthread_tls_t entirely?

Comment (mppf) [inline runtime/include/tasks/qthreads/tasks-qthreads.h]: Since qthreads doesn't (apparently) give a way to get the arguments provided when starting the task, we still have to stash a pointer to the task bundle. I'd expect it is clearest code if chpl_qthread_tls_t continues to have that pointer in it. Then, it's just a matter of whether or not that type has prvdata, lock_filename, and lock_lineno; or if those are stored in the argument bundle (even though they are never used to pass data).

Comment (gbtitus) [inline runtime/include/tasks/qthreads/tasks-qthreads.h]: Can we not get rid of chpl_qthread_tls_t and store only the address of the bundle Qthread task-local space?

Comment (mppf) [inline runtime/include/tasks/qthreads/tasks-qthreads.h]: I think the code would be clearest if chpl_qthread_tls_t contains just that pointer. But then, yes, we'd be storing only a pointer in the qthread task-local space (instead of a pointer and 3 integers).

Comment (gbtitus) [inline runtime/include/tasks/qthreads/tasks-qthreads.h]: I'm failing to understand the need for chpl_qthread_tls_t at all.  Why can we not say that qthread_get_tasklocal() returns the address of a chpl_task_bundle_t*, rather than that it returns the address of a chpl_qthread_tls_t struct that contains as its only member a chpl_task_bundle_t*? What does the chpl_qthread_tls_t container get us?

**Comment by ronawho** [comment]
@mppf Do you have any sense of when you might get back to this?

**Comment by mppf** [comment]
@ronawho - I hope to look at it in December. We were so close, last time...

**Comment by ronawho** [comment]
Ok, thanks. 

Just FYI: @daviditen and I are looking at it this WRT to task spawning times since that's a large cause of the LCALS performance loss compared to OpenMP. This PR has a non-trivial impact on task startup cost:


```
# parallel/taskCompare/elliot/empty-forall
#
# Date  Elapsed time:

# default
11/20/16        21.3854
11/20/16        21.474
11/20/16        21.2531

# spinwait
11/20/16        5.97304
11/20/16        5.96263
11/20/16        5.99871

# spinwait + stack arg bundles (#4074)
11/20/16        5.24702
11/20/16        5.32391
11/20/16        5.27926
```

Reducing task startup time (with a spinwait in qthreads) has had a pretty dramatic impact on a lot of perf tests, so it's likely this will have a similar improvement on [those tests](http://chapel.sourceforge.net/perf/chapcs/taskSpawn/?startdate=2016/11/12&enddate=2016/11/22&graphs=ssca2kernels23scale8,ssca2kernel4scale8,lcalsrawompshort,lcalsrawompmedium,lcalsrawomplong,ssca2scale8,luleshrelease,hpcchplperformancegflops,ssca2kernel4tepsscale8,luleshbradcstudyversion,minimdljsize10time,parboilhistoserialexecutiontime,nasparallelbenchmarkscgtimingssizew,nasparallelbenchmarksfttimingssizea,2darrayassignment1024x1024,2dnonlocalarrayassignment1024x1024,arrayinitialization,emptytaskspawntimings500000xmaxtaskpar,nasparallelbenchmarkseptimingssizea,hpcchpltime,reductionstimesec,timetobuildtrees).

File: compiler/passes/parallel.cpp [hunk 16]
```diff
@@ -341,8 +351,18 @@ bundleArgs(CallExpr* fcall, BundleArgsFnData &baData) {
...
-                      ctype, newMemDesc("bundled args"));
+
+  // We could potentially heap-allocate based upon the size of the
+  // argument bundle, but the local case would pass these arguments
+  // on the stack anway, so it seems moot.
+
+  fcall->insertBefore( new CallExpr(PRIM_MOVE,
+                          tempc,
+                          new CallExpr(PRIM_STACK_ALLOCATE_CLASS,
+                            ctype->symbol)));
```
Comment (ronawho) [inline compiler/passes/parallel.cpp]: Minor, but the space formatting for this this call is hard to read

File: compiler/passes/parallel.cpp [hunk 17]
```diff
@@ -659,43 +679,52 @@ static void create_block_fn_wrapper(FnSymbol* fn, CallExpr* fcall, BundleArgsFnD
...
-    // Special case: 
-    // If this is an on block, remember the first field,
-    // but don't add to the list of actuals passed to the original on_fn.
-    // It contains the locale on which the new task is launched.
-    if (first && fn->hasFlag(FLAG_ON))
-      /* no-op */;
-    else
-      call_orig->insertAtTail(tmp);
+    // Skip runtime header
+    if (i > 0) {
```
Comment (ronawho) [inline compiler/passes/parallel.cpp]: Seems like a weird way to do this. Can this just be something like:

```C
if (field == ctype->fields.first) continue;
```

File: runtime/src/comm/gasnet/comm-gasnet.c [hunk 18]
```diff
@@ -1302,85 +1562,28 @@ void  chpl_comm_execute_on(c_nodeid_t node, c_sublocid_t subloc,
...
-                                 CHPL_RT_MD_COMM_FRK_SND_ARG, 0, 0);
-    chpl_memcpy(argCopy, arg, arg_size);
-    *(void**)(&(info->arg)) = argCopy;
-  }
+    assert(0); // locale model code should prevent this...
 
-  if (chpl_nodeID == node) {
-    if (info->serial_state)
-      fork_nb_wrapper(info);
+    // Visual Debug?  Should we generate a task here???
```
Comment (gbtitus) [inline runtime/src/comm/gasnet/comm-gasnet.c]: Won't we do the appropriate vdebug task callback in the tasking layer, if we make a task?  If we don't make a task here (i.e. if we do an ftable_call), we definitely should not do a vdebug task callback, that's certain.

**Comment by gbtitus** [comment]
Looks good -- okay to merge as far as the module and runtime code goes.

File: runtime/include/tasks/fifo/tasks-fifo.h [hunk 19]
```diff
@@ -24,11 +24,6 @@
 
 #include "chpl-threads.h"
 
-#ifdef __cplusplus
-extern "C" {
-#endif
```
Comment (mppf) [inline runtime/include/tasks/fifo/tasks-fifo.h]: I'm guessing this is a merge error - thanks for pointing it out - I'll un-remove it

File: compiler/passes/parallel.cpp (hunk 17 above)
Comment (mppf) [inline compiler/passes/parallel.cpp]: that doesn't work well because fields.first is an Expr* while field is a Symbol*. Besides that, the general pattern of if ( something ) continue; here is challenging because the i++ still needs to happen (for the next loop anyway). I think the conditional is actually the simplest way to write the correct solution.
